Docs: Add ViewState Documentation

paulhdk opened the pull request:
# Docs: Add ViewState Documentation

## :recycle: Current situation & Problem

As pointed out in https://github.com/StanfordSpezi/SpeziLLM/pull/61#discussion_r1720129130 , `ViewState` was lacking documentation.

## :gear: Release Notes 
* Add `ViewState` documentation.


## :books: Documentation
.


## :white_check_mark: Testing
.


## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

PSchmiedmayer commented on Sources/SpeziViews/Model/ViewState.swift:
```diff
@@ -10,6 +10,52 @@ import Foundation
+/// import SpeziViews
+/// import SwiftUI
+///
+/// enum ProcessingError: LocalizedError {
+///     case resourceError
+/// }
+///
+/// func doSomeProcessing() throws {
+///     // Function may throw an error
+///     throw ProcessingError.resourceError
+/// }
```
I would suggest to remove these types here and just add a `// ...` placeholder in the button definition that entails that it could call a throwing and async function in here.

PSchmiedmayer commented on Sources/SpeziViews/Model/ViewState.swift:
```diff
@@ -10,6 +10,52 @@ import Foundation
+///     throw ProcessingError.resourceError
+/// }
+///
+/// struct OperationStateTestView: View {
```
```suggestion
/// struct ViewStateExample: View {
```

PSchmiedmayer commented on Sources/SpeziViews/Model/ViewState.swift:
```diff
@@ -10,6 +10,52 @@ import Foundation
+///                 do {
+///                     try doSomeProcessing()
+///                     viewState = .idle
+///                 } catch let error as LocalizedError {
+///                     viewState = .error(error)
+///                 } catch {
+///                     print("An unknown error occurred: \(error)")
+///                 }
```
I think it is not a good example to show a log that is never processed; I would suggest to point to `AnyLocalizedError` as a nice way to catch localized errors. And a good way to point to this in a `Tip` above when we might cover that error takes a `LocalizedError`?

paulhdk replied:
Yes, I wasn't quite happy with that simple `print` either, but leaving out the last `catch` will throw a compiler warning, won't it? Because not every error is a `LocalizedError`. Is there a neat way to avoid that?

paulhdk replied:
Oh, nevermind, that's what `AnyLocalizedError` is for. My bad! :)

PSchmiedmayer commented on Sources/SpeziViews/Model/ViewState.swift:
```diff
@@ -10,6 +10,52 @@ import Foundation
 /// Allows SwiftUI views to keep track of their state and communicate to outside views.
+///
+/// A `ViewState` provides a built-in mechanism for tracking the state of a Spezi UI component.
+/// A view can be in either an `idle`, `processing`, or `error` `ViewState`.
```
I would suggest to note that a `ViewState` takes a `LocalizedError` as an associated value and would point to the Apple Documentation on this.

We should also add a small tip to our "best effort" matching of non-Localized errors in `AnyLocalizedError` in this package.

PSchmiedmayer commented on Sources/SpeziViews/Model/ViewState.swift:
````diff
@@ -10,6 +10,52 @@ import Foundation
+/// }
+/// ```
+///
+/// The ``SwiftUI/View/viewStateAlert(state:)-4wzs4`` view modifier can be used to automatically notify users when the
````
```suggestion
/// The ``SwiftUI/View/viewStateAlert(state:)-4wzs4`` view modifier can be used to automatically notify users with an [`Aler`](https://developer.apple.com/documentation/swiftui/view/alert(_:ispresented:actions:)-3npin) when the
```

PSchmiedmayer commented on Sources/SpeziViews/Model/ViewState.swift:
```diff
@@ -10,6 +10,52 @@ import Foundation
+/// `ViewState` enters an error state.
+///
+/// > Tip:
+/// > To avoid having to handle state changes manually (as seen above), a `ViewState` can be mapped to an
+/// > `OperationState`, allowing the `ViewState` to change automatically when the application's internal state changes.
```
I am not sure if this really expresses what `OperationState` does. Operation state rather allows you to define more custom cases that would be relevant to your application and then maps it back to `ViewState` to make it easy to use the same `viewStateAlert` modifiers and other elements. We might want to rephrase this a bit.

PSchmiedmayer reviewed (changes requested):
Thank you for working on the documentation @paulhdk; greatly appreciated!

I added some comments, let me know if you have any follow-up questions. You should also have write access to all Spezi repos to create a branch right in the repo so the CodeCov checks are passing with your next PR.

paulhdk commented:
> Thank you for working on the documentation @paulhdk; greatly appreciated!
> 
> I added some comments, let me know if you have any follow-up questions. You should also have write access to all Spezi repos to create a branch right in the repo so the CodeCov checks are passing with your next PR.

I initially tried pushing a branch directly into the SpeziViews repo, but wasn't able to do so. There might have also been something wrong on my end, so I'll do some more investigating for my next PR.

Thank you for all your comments, @PSchmiedmayer! I've addressed them all. PTAL.
